library/x509_crt.c

```diff
@@ -3005,6 +3005,127 @@ static int x509_crt_verify_chain(
+#endif
+#else
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#endif
+
+#ifndef AF_INET6
+
+/* definition located further below to possibly reduce compiler inlining */
+static int x509_inet_pton_ipv4( const char *src, void *dst );
+
+#define li_cton(c,n) \
+  (((n) = (c) - '0') <= 9 || (((n) = ((c)&0xdf) - 'A') <= 5 ? ((n) += 10) : 0))
+
+static int x509_inet_pton_ipv6( const char *src, void *dst )
+{
+    const unsigned char *v = (const unsigned char *)src;
+    int i = 0, j, dc = -1;
+    uint16_t addr[8];
+    do
+    {
+        /* note: allows excess leading 0's, e.g. 1:0002:3:... */
+        uint16_t x = j = 0;
+        for( uint8_t n; j < 4 && li_cton( *v, n ); x <<= 4, x |= n, ++v, ++j ) ;
+        if( j != 0 )
+        {
+            addr[i++] = ( x << 8 ) | ( x >> 8 ); /* htons(x) */
+            if( *v == '\0' )
+                break;
+            else if( *v == '.' && ( i != 0 || dc != -1 ) && ( i < 7 ) &&
+                     /* walk back to prior ':', then parse as IPv4-mapped */
+                     ( *--v == ':' || *--v == ':' ||
+                       *--v == ':' || *--v == ':' ) &&
+                     x509_inet_pton_ipv4( (const char *)++v, addr+i ) == 0 )
+            {
+                i += 2;
+                break;
+            }
+            else if( *v != ':' )
+                return( -1 );
+        }
+        else
+        {
+            if( dc != -1 || *v != ':' || ( ( dc = i ) == 0 && *++v != ':' ) )
+                return( -1 );
+            if( v[1] == '\0' )
+                break;
+        }
+        ++v;
+    } while( i < 8 );
+    if( i > ( dc != -1 ? 6 : 8 ) )
+        return( -1 );
+
+    if( dc != -1 )
+    {
+        i -= dc;
+        j = 8 - i - dc;
+        if( i )
+            memmove( addr + dc + j, addr + dc, i * sizeof(*addr) );
+        memset( addr + dc, 0, j * sizeof(*addr) );
+    }
+    memcpy( dst, addr, sizeof(addr) );
+    return( 0 );
+}
+
+static int x509_inet_pton_ipv4( const char *src, void *dst )
+{
+    /* note: allows leading 0's, e.g. 000.000.000.000 */
+    const unsigned char *v = (const unsigned char *)src;
+    uint8_t *res = (uint8_t *)dst;
+    uint8_t d1, d2, d3, i = 0;
+    const uint8_t tens[] = { 0, 10, 20, 30, 40, 50, 60, 70, 80, 90 };
+    do
+    {
+        if (      ( d1 = *(uint8_t *)  v - '0' ) > 9 )
+            break;
+        else if ( ( d2 = *(uint8_t *)++v - '0' ) > 9 )
+            *res++ = d1;
+        else if ( ( d3 = *(uint8_t *)++v - '0' ) > 9 )
+            *res++ = tens[d1] + d2;
+        else if ( d1 == 0 )
+            *res++ =   0 + tens[d2] + d3, ++v;
+        else if ( d1 == 1 )
+            *res++ = 100 + tens[d2] + d3, ++v;
+        else if ( d1 == 2 )
+            *res++ = 200 + tens[d2] + d3, ++v;
+    } while( ++i < 4 && *v++ == '.' );
+    return( i == 4 && *v == '\0' ? 0 : -1 );
+}
+
+#else
+
+static int x509_inet_pton_ipv6( const char *src, void *dst )
+{
+    return( inet_pton( AF_INET6, src, dst ) == 1 ? 0 : -1 );
```

Comment:
How confident are we that `inet_pton` is consistent across platforms? I'm worried that some embedded platforms might have a limited `inet_pton` implementation that's good enough for some specific use case and causes mysterious problems when used in a certificate deployment.

Reply:
The tests that I will add (later this evening) to the mbedtls test suite is the best way to address these concerns, both for the platform `inet_pton()` or the local implementation.

Reply:
I have added some 29 tests for the code which calls `inet_pton()` or the local implementation.  Of note, my local implementation supports leading 0's in IPv4 octets (up to 3 digits per octet) (e.g. `010.010.010.010`), but `inet_pton()` on my Linux system rejects that.
:rocket:

Reply:
Linux inet_pton() accepts IPv6 "00000::0" while my local implementation rejects more than 4 hex digits per IPv6 address segment.